File: architecture-skills/design-patterns/examples/observer/observer.cpp
// Observer pattern in C++.
//
// std::function is the modern C++ observer: any lambda, functor, or
// bound member can subscribe without an interface hierarchy. subscribe()
// returns an id so callers can unsubscribe (avoiding the classic
// dangling-observer bug).

#include <functional>
#include <iostream>
#include <string>
#include <vector>

using PriceObserver = std::function<void(const std::string&, double)>;

class StockTicker {  // subject
public:
    int subscribe(PriceObserver observer) {
        observers_.push_back({next_id_, std::move(observer)});
        return next_id_++;
    }

    void unsubscribe(int id) {
        std::erase_if(observers_, [id](const auto& entry) { return entry.first == id; });
    }

    void updatePrice(const std::string& symbol, double price) {
        for (const auto& [id, observer] : observers_) {
            observer(symbol, price);
        }
    }

private:
    std::vector<std::pair<int, PriceObserver>> observers_;
    int next_id_ = 0;
};

class PriceDisplay {
public:
    explicit PriceDisplay(std::string name) : name_(std::move(name)) {}

    void onPrice(const std::string& symbol, double price) const {
        std::cout << name_ << ": " << symbol << " is now $" << price << '\n';
    }

private:
    std::string name_;
};

int main() {
    std::cout.precision(2);
    std::cout << std::fixed;

    StockTicker ticker;
    PriceDisplay display("Lobby display");

    int displayId = ticker.subscribe(
        [&display](const std::string& symbol, double price) {
            display.onPrice(symbol, price);
        });

    ticker.subscribe([](const std::string& symbol, double price) {
        if (symbol == "ACME" && price > 100.0) {
            std::cout << "ALERT: ACME crossed $100.00 (now $" << price << ")\n";
        }
    });

    ticker.updatePrice("ACME", 98.5);
    ticker.updatePrice("ACME", 101.25);

    ticker.unsubscribe(displayId);
    ticker.updatePrice("GLOBEX", 45.0);  // only the alert remains
    return 0;
}

File: architecture-skills/design-patterns/examples/observer/observer.py
"""Observer pattern in Python.

A stock ticker notifies subscribers of price changes. Plain callables
are the idiomatic observer in Python — any function, lambda, or bound
method can subscribe, no interface required.
"""

from __future__ import annotations

from typing import Callable

PriceObserver = Callable[[str, float], None]


class StockTicker:
    """Subject: keeps prices and notifies subscribers on change."""

    def __init__(self) -> None:
        self._prices: dict[str, float] = {}
        self._observers: list[PriceObserver] = []

    def subscribe(self, observer: PriceObserver) -> Callable[[], None]:
        self._observers.append(observer)

        def unsubscribe() -> None:
            self._observers.remove(observer)

        return unsubscribe

    def update_price(self, symbol: str, price: float) -> None:
        self._prices[symbol] = price
        for observer in list(self._observers):
            observer(symbol, price)


class PriceDisplay:
    """An observer implemented as a class with a bound method."""

    def __init__(self, name: str) -> None:
        self.name = name

    def on_price(self, symbol: str, price: float) -> None:
        print(f"{self.name}: {symbol} is now ${price:.2f}")


def make_alert(symbol: str, threshold: float) -> PriceObserver:
    """An observer implemented as a closure."""

    def on_price(updated_symbol: str, price: float) -> None:
        if updated_symbol == symbol and price > threshold:
            print(f"ALERT: {symbol} crossed ${threshold:.2f} (now ${price:.2f})")

    return on_price


if __name__ == "__main__":
    ticker = StockTicker()
    lobby_display = PriceDisplay("Lobby display")
    unsubscribe = ticker.subscribe(lobby_display.on_price)
    ticker.subscribe(make_alert("ACME", 100.0))

    ticker.update_price("ACME", 98.5)
    ticker.update_price("ACME", 101.25)

    unsubscribe()
    ticker.update_price("GLOBEX", 45.0)  # only the alert remains subscribed
